### PR TITLE
Report an error for an unsupported type tests involving 'ref' structs.

### LINKED ISSUE #63085: Type test for ref struct is incorrect
If `ref struct G<T>` defined, `this` type test inside the struct for any resolved generic returns false (and compiler not even add test in IL).
It also shows warning CS0184 and does not allow to use it in pattern math type test.

**Version Used**:  C# 10

**Steps to Reproduce**:

Compile and run next code:

Use next code:

```csharp
using System;

new G<int>().Test();
new G<object>().Test();

ref struct G<T>
{
    public void Test()
    {
        if (this is G<int>)
        {
            Console.WriteLine("int");
        }
        else if (this is G<object>)
        {
            Console.WriteLine("object");
        }
        else
        {
            Console.WriteLine("unknown");
            Console.WriteLine(typeof(T));
        }
    }

    /*public void TestPattern()
    {
        var genericTypePattern = this switch
        {
            G<int> => "int",
            G<object> => "object",
            _ => "unknown"
        };

        Console.WriteLine(genericTypePattern);
    }*/
}
```
[On SharpLib.IO](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoavGAdwAIBxAHgEs8MA+ACgEoAdABUYAZwwCA3LQYtWEYACsYYHgJHjJ/GTSqwAZowm4184d2oBvaozuMiAZgcpGoiQNv2bVe38bsRrwYABbsYgERbJw8/F7+jD4JCURkAJy8AEQxmTrx/gC++X4wADZiMAFBoeGR8ooqatxxvslJyX6pGZkNqhi5uh1FrQllFcXeEynpWfgA1ngQ9HgDU/5dwQCeAA4wEAa8wvx5I4X5w/kA9ABUTi5uWgAKAIYYGDBQeJ6n7QkAbs8oIwAOYwOhQdhgYQ7GAvN4fPCMAC8jBqETE9HYGDAITWvw68hi3GRxOyXEyaDWfjYvSaJMYPWUfQpVPsAH16Zl5otlpk1gVdGsNqDwZDobs4e9Pic/AVrpdqAUgA==)

**Expected Behavior**:
Program should output:
```
int
object
```

No CS0184 warning should not be shown, if uncomment, TestPattern method should compile without errors.

**Actual Behavior**:
Program outputs:
```
unknown
System.Int32
unknown
System.Object
```

CS0184 warning shown, TestPattern method can not be compiled without changes

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #63085.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler, @jaredpar Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler, @jaredpar Please review

**Review by jaredpar — APPROVED** [review]

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md [hunk 1]
```diff
@@ -1,5 +1,22 @@
 # This document lists known breaking changes in Roslyn after .NET 6 all the way to .NET 7.
 
+## Type tests for `ref` structs ar not supported.
+
+***Introduced in Visual Studio 2022 version 17.4***
+
+When a `ref` struct type is used in an 'is' or 'as' operator, in some scenarios compiler was previously reporting
+an erroneous warning about the type test always failing at runtime, omitting the actual type check, and leading to
+an incorrect behavior. When an incorrect behavior at execution time was possible, compiler will now produce an error instead.
```
Comment (333fred) [inline docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:9]: `an` should be removed in both places on this line.

**Review by 333fred — APPROVED** [review]
LGTM (commit 2). Small grammar correction in the breaking changes.

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md (hunk 1 above)
Comment (AlekseyTs) [reply docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:9]: > ```an``` should be removed in both places on this line.

Could you clarify please? There are three places where ```an``` is used on this line.

Comment (333fred) [reply docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:9]: Sorry, yes.
```suggestion
incorrect behavior. When incorrect behavior at execution time was possible, compiler will now produce an error instead.
```

**Pull request merged**